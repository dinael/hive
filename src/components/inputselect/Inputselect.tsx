import { FC, ChangeEvent } from 'react'

import styles from './Inputselect.module.scss'

import { Icon } from '@/components/icon/Icon'
import { ValidationMessage } from '@/components/validationmessage/ValidationMessage'
import { Label as LabelText } from '@/components/label/Label'

export type Option = {
  id?: string
  value?: string
  disabled?: boolean
  hidden?: boolean
  select?: boolean
  option: string
}

export type InputselectProps = {
  label: string
  labelHidden?: boolean
  id?: string
  name?: string
  value: string
  onChange: (event: ChangeEvent<HTMLSelectElement>) => void
  options: Option[] | { [groupName: string]: Option[] }
  error?: boolean
  errorText?: string
  disabled?: boolean
  required?: boolean
  isFocused?: boolean
  className?: string
}

export const Inputselect: FC<InputselectProps> = ({
  label,
  labelHidden = false,
  id,
  name,
  value,
  onChange,
  options,
  error,
  errorText,
  disabled,
  required = false,
  isFocused,
  className,
  ...props
}) => {

  const selectStyles = `
    ${styles['input-select-component']}
    ${error ? 'has-error' : ''}
    ${className || ''}
  `

  const renderOptions = (options: InputselectProps['options']) => {
    if (Array.isArray(options)) {
      return options.map((option) => (
        <option
          key={option.id}
          value={option.value}
          disabled={option.disabled}
          hidden={option.hidden}
          selected={option.select}>
          {option.option}
        </option>
      ))
    } else {
      return Object.entries(options).map(([groupName, groupOptions]) => (
        <optgroup key={groupName} label={groupName}>
          {groupOptions.map((option) => (
            <option
              key={option.id}
              value={option.value}
              disabled={option.disabled}
              hidden={option.hidden}
              selected={option.select}>
              {option.option}
            </option>
          ))}
        </optgroup>
      ))
    }
  }

  return (
    <label
      className={selectStyles}
      htmlFor={id}
      {...props}>
      <LabelText
        onlyText
        className={`${styles['input-select-label']}`}
        text={label}
        required={required}
        isRequiredText={'Este elemento es requerido'}
        labelHidden={labelHidden} />
      <select
        name={name}
        id={id}
        value={value}
        onChange={onChange}
        required={required}
        disabled={disabled}
        aria-describedby={`validation-message-${id}`}>
        {renderOptions(options)}
      </select>
      <Icon
        className={styles.icon}
        name="chevron-down"
        size="xs"
        title={''} />
      {error && !isFocused ?
        <ValidationMessage
          className={styles['input-select-validation-message']}
          message={`${errorText}`}
          kind='error'
          role="alert"
          aria-live="assertive"
          aria-atomic="true"
          id={`validation-message-${id}`} />
        : null}
    </label>
  )
}

export default Inputselect
