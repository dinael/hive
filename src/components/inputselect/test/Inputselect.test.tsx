import { describe, test, expect } from 'vitest'
import { render, screen } from '@testing-library/react'
import { Inputselect, InputselectProps } from '../Inputselect'

const options = [
  { option: 'Opción 1', value: '1' },
  { option: 'Opción 2', value: '2' },
  { option: 'Opción 3', value: '3' },
]

const groupOptions = {
  'Grupo 1': [
    { option: 'Opción 4', value: '4' },
    { option: 'Opción 5', value: '5' },
  ],
  'Grupo 2': [
    { option: 'Opción 6', value: '6' },
    { option: 'Opción 7', value: '7' },
  ],
}

describe('Inputselect', () => {
  const defaultProps: InputselectProps = {
    label: 'label-test',
    labelHidden: false,
    id: 'id-test',
    name: 'name-test',
    disabled: false,
    options: options,
    error: false,
    errorText: 'errorText-test',
    required: false,
    isFocused: false,
    className: 'className-test'
  }

  test('Inputselect renders snapshot', () => {
    const { container } = render(<Inputselect {...defaultProps} />)
    expect(container).toMatchSnapshot()
  })

  test('Inputselect renders snapshot with group options', () => {
    const { container } = render(<Inputselect {...defaultProps} options={groupOptions} />)
    expect(container).toMatchSnapshot()
  })
})
