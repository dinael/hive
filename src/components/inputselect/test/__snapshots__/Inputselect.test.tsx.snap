// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Inputselect > renders snapshot 1`] = `
<div>
  <label
    class="
        inputselect-component
        
        className-test
      "
    for="id-test"
  >
    <span
      class="
      label-text-component
      
      
      
      inputselect-label
      "
    >
      label-test:
    </span>
    <select
      aria-describedby="validation-message-id-test"
      id="id-test"
      name="name-test"
    >
      <option
        value="1"
      >
        Opción 1
      </option>
      <option
        value="2"
      >
        Opción 2
      </option>
      <option
        value="3"
      >
        Opción 3
      </option>
    </select>
    <span
      aria-hidden="true"
      class="
        icon-component
        xs
        icon
      "
      style="--icon-color: currentColor;"
    >
      <svg
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
        xmlns:xlink="http://www.w3.org/1999/xlink"
      >
        <title>
          chevron-down
        </title>
        <use
          xlink:href="/src/assets/sprite.svg#chevron-down"
        />
      </svg>
    </span>
  </label>
</div>
`;

exports[`Inputselect > renders snapshot with group options 1`] = `
<div>
  <label
    class="
        inputselect-component
        
        className-test
      "
    for="id-test"
  >
    <span
      class="
      label-text-component
      
      
      
      inputselect-label
      "
    >
      label-test:
    </span>
    <select
      aria-describedby="validation-message-id-test"
      id="id-test"
      name="name-test"
    >
      <optgroup
        label="Grupo 1"
      >
        <option
          value="4"
        >
          Opción 4
        </option>
        <option
          value="5"
        >
          Opción 5
        </option>
      </optgroup>
      <optgroup
        label="Grupo 2"
      >
        <option
          value="6"
        >
          Opción 6
        </option>
        <option
          value="7"
        >
          Opción 7
        </option>
      </optgroup>
    </select>
    <span
      aria-hidden="true"
      class="
        icon-component
        xs
        icon
      "
      style="--icon-color: currentColor;"
    >
      <svg
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
        xmlns:xlink="http://www.w3.org/1999/xlink"
      >
        <title>
          chevron-down
        </title>
        <use
          xlink:href="/src/assets/sprite.svg#chevron-down"
        />
      </svg>
    </span>
  </label>
</div>
`;
